### PR TITLE
Blog post about code coverage and mutation testing

### PR DESCRIPTION
**Pull request by rachelcarmena**

### CODE REVIEW

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+image:
+  src: /assets/img/custom/blog/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.jpg
+tags:
+- coverage
+- testing
+- mutation testing
+- quality
+
+---
+A long time ago, I met a development team which was working under big pressure by the quality team. Personally, I don’t like this kind of difference between development and quality teams, because it makes development teams feel that they aren’t responsible of quality and it provokes confrontations between them instead of working in a collaborative environment.
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: `I don’t like this kind of difference**s** between`

Comment (ghost) [reply _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: `it makes development teams feel that they aren’t responsible of quality and`
I propose: it leads to development teams not feeling responsible **for** quality
` it provokes confrontations between them instead of working in a collaborative environment.`
I suggest: it leads to a confrontational relationship where they should be working collaboratively towards a unified goal of delivering a quality product.

**Review by ghost — COMMENTED** [review]
need to continue review later

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- testing
+- mutation testing
+- quality
+
+---
+A long time ago, I met a development team which was working under big pressure by the quality team. Personally, I don’t like this kind of differences between development and quality teams, because it leads to development teams not feeling responsible for quality and to a confrontational relationship. They should be working collaboratively towards a unified goal of delivering a quality product.
+
+One of the requirements was to have more than 85% of code coverage to ensure code quality. The result was perverse: development team wrote tests without assertions; they only invoked methods with different arguments to reach the desired percentage. It’s clear that they didn’t follow TDD.
+
+Code coverage only gives us information about the percentage of code lines which are executed during tests. Let’s see a way to check if we have a good safety net with current tests to make changes in production code.
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: I would put the first line closer to the previous paragraph, `Code coverage only gives us information about the percentage of code lines which are executed during tests.`;
and start the other paragraph with the second line.
`Let’s see a way to check if we have a good safety net with current tests to make changes in production code.`

Comment (ghost) [reply _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: Also I would rephrase the second line:
`Let’s see a way to check if we have a good safety net with current tests to make changes in production code.`
with something like this:
Let's consider another way to verify that our current tests provide us with a safety net when we make changes to production code.

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+If we made a change in production code, such as “replacing + with -”, “replacing < with >=” or “returning a different value”, test battery might detect that change. In order words, tests might fail.
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: If for example we change the production code; and swap a `+` with a `-` sign or replace `<` with `>=`; or we return a different value in a given method: tests should detect that change; and they should fail.

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+There are tools to make changes in production code automatically and to run tests in order to check if those changes are detected. It’s talked about:
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: `It’s talked about:`
I would write: 
It is usually referred to as follow

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+</center>
+<br/>
+
+And I imagined a situation such as this:
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/super-pang-mutation-testing.png" alt="Super Pang - Mutation Testing" class="img img-responsive"/>
+</center>
+
+It’s called **mutation testing** and it's a good way to make sure that you have a good safety net with your current tests to refactor production code or to add new features, among others.
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: `among others.`?
You mean among other things?
You mean adding a feature to existing set of features?

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+```
+
+In that case, we can succeed in killing every mutation:
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/new-pit-statistics.png" alt="PIT Statistics" class="img img-responsive"/>
+</center>
+<br/>
+
+Others prefer <a href="https://projectlombok.org/features/EqualsAndHashCode" target="_blank">Lombok</a> but think about it whether you only need to compare objects.
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: `but think about it whether you only need to compare objects.`
Not sure I understand, you mean 
think about it _only_ when you need to compare objects. ?
Or
Think about using Lombok if the only thing you need is to compare object?

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+Regarding _verification_, <a href="https://static.javadoc.io/org.mockito/mockito-core/2.8.47/org/mockito/ArgumentMatchers.html#refEq(T,%20java.lang.String...)">`refEq`</a> is available from <a href="http://site.mockito.org" target="_blank">Mockito</a>.
+
+## Conclusion
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: Not sure we need to say that it is the conclusion, but this a matter of style I suppose. So don''t take my word for it; just my liking here :)

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+We have seen that code coverage doesn’t ensure that we have a good safety net to change production code. And mutation testing can be useful to test your tests in order to get more information about their suitability.
+
+Take a look to <a href="/2014/12/14/quality-cannot-be-measured/">_"Code quality cannot be measured"_</a> by **Sandro Mancuso**.
```
Comment (ghost) [inline _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: Take a look **at** [..] for more on this topic.

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md (hunk 5 above)
Comment (rachelcarmena) [reply _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: `among other things`, but I can remove it. It's not necessary. Thank you @halimakoundi !

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md (hunk 6 above)
Comment (rachelcarmena) [reply _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: Lombok is not necessary if the only thing you need is to compare objects. Thanks @halimakoundi !!

File: _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md (hunk 8 above)
Comment (rachelcarmena) [reply _posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md]: Thank you very much for your help!! @halimakoundi, you're awesome!!!! And a GREAT colleague!!!

**Comment by rachelcarmena** [comment]
It's ready for merging. Thanks again @halimakoundi !!
